generator/feature_segments_checker/feature_segments_checker.cpp

```diff
@@ -212,20 +239,37 @@ class Processor
     if (realFeatureLengthMeters == 0.0)
       return;
 
-    double distFromStartMeters = 0;
+    double const k = (endAltitude - startAltitude) / realFeatureLengthMeters;
     for (uint32_t i = 1; i + 1 < numPoints; ++i)
     {
-      // Feature segment length.
-      double const segmentLengthMeters =
-          MercatorBounds::DistanceOnEarth(f.GetPoint(i - 1), f.GetPoint(i));
-      distFromStartMeters += segmentLengthMeters;
-
-      generator::SrtmTile::THeight const pointAltitude =
-          m_srtmManager.GetHeight(MercatorBounds::ToLatLon(f.GetPoint(i)));
-      int32_t const deviation = static_cast<int32_t>(ExpectedPointAltitude(startAltitude, endAltitude, distFromStartMeters,
-                                                                           realFeatureLengthMeters)) - pointAltitude;
+      int32_t const deviation =
+          static_cast<int32_t>(ExpectedPointAltitude(k, startAltitude, pointDists[i])) - pointAltitudes[i];
       m_diffFromLinear[deviation]++;
     }
+
+    // Linear least squares for feature points.
+    {
+      long double const distSum = accumulate(pointDists.begin(), pointDists.end(), 0.0);
+      uint64_t const altSum = accumulate(pointAltitudes.begin(), pointAltitudes.end(), 0);
+      long double const distSquareSum = accumulate(pointDists.begin(), pointDists.end(), 0.0,
+                                                   [](double x, double y) { return x + y * y; });
+      long double distMultiplyAltSum = 0;
+      for (size_t i = 0; i < pointDists.size(); ++i)
+        distMultiplyAltSum += pointDists[i] * pointAltitudes[i];
+
+      if (distSum == 0.0 || distSquareSum == 0.0)
+        return;
+
+      double const b = (altSum/distSum - distMultiplyAltSum/distSquareSum) / (numPoints/distSum - distSum/distSquareSum);
```

Comment:
Below the correct formulas if to calculate k first:
```c++
k = (altSum/numPoints - distMultiplyAltSum/distSum) / (distSum/numPoints - distSquareSum/distSum);
b = (altSum - k * distSum) / numPoints;
```
But if to do do the entropy would be a little more. I rewrote the code in another way. The entropy is a little less this way. 

> n any case, your expression for b is quite complex and may be simplified.

Yes it's possible. But in that case we'll have too big values while calculation which effects the precision of the result.